File: src/generator/AutoRest.Java.Tests/src/test/java/fixtures/modelflattening/ModelFlatteningTests.java
package fixtures.modelflattening;

import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import fixtures.modelflattening.implementation.AutoRestResourceFlatteningTestServiceImpl;
import fixtures.modelflattening.models.FlattenParameterGroup;
import fixtures.modelflattening.models.FlattenedProduct;
import fixtures.modelflattening.models.Resource;
import fixtures.modelflattening.models.ResourceCollection;
import fixtures.modelflattening.models.SimpleProduct;

public class ModelFlatteningTests {
    private static AutoRestResourceFlatteningTestService client;

    @BeforeClass
    public static void setup() {
        client = new AutoRestResourceFlatteningTestServiceImpl("http://localhost:3000");
    }

    @Test
    public void getArray() throws Exception {
        List<FlattenedProduct> result = client.getArray();
        Assert.assertEquals(3, result.size());
        // Resource 1
        Assert.assertEquals("1", result.get(0).id());
        Assert.assertEquals("OK", result.get(0).provisioningStateValues());
        Assert.assertEquals("Product1", result.get(0).pname());
        Assert.assertEquals("Flat", result.get(0).flattenedProductType());
        Assert.assertEquals("Building 44", result.get(0).location());
        Assert.assertEquals("Resource1", result.get(0).name());
        Assert.assertEquals("Succeeded", result.get(0).provisioningState());
        Assert.assertEquals("Microsoft.Web/sites", result.get(0).type());
        Assert.assertEquals("value1", result.get(0).tags().get("tag1"));
        Assert.assertEquals("value3", result.get(0).tags().get("tag2"));
        // Resource 2
        Assert.assertEquals("2", result.get(1).id());
        Assert.assertEquals("Resource2", result.get(1).name());
        Assert.assertEquals("Building 44", result.get(1).location());
        // Resource 3
        Assert.assertEquals("3", result.get(2).id());
        Assert.assertEquals("Resource3", result.get(2).name());
    }

    @Test
    public void putArray() throws Exception {
        List<Resource> body = new ArrayList<>();
        FlattenedProduct product = new FlattenedProduct();
        product.withLocation("West US");
        product.withTags(new HashMap<String, String>());
        product.tags().put("tag1", "value1");
        product.tags().put("tag2", "value3");
        body.add(product);
        FlattenedProduct product1 = new FlattenedProduct();
        product1.withLocation("Building 44");
        body.add(product1);
        client.putArray(body);
    }

    @Test
    public void getDictionary() throws Exception {
        Map<String, FlattenedProduct> result = client.getDictionary();
        Assert.assertEquals(3, result.size());
        // Resource 1
        Assert.assertEquals("1", result.get("Product1").id());
        Assert.assertEquals("OK", result.get("Product1").provisioningStateValues());
        Assert.assertEquals("Product1", result.get("Product1").pname());
        Assert.assertEquals("Flat", result.get("Product1").flattenedProductType());
        Assert.assertEquals("Building 44", result.get("Product1").location());
        Assert.assertEquals("Resource1", result.get("Product1").name());
        Assert.assertEquals("Succeeded", result.get("Product1").provisioningState());
        Assert.assertEquals("Microsoft.Web/sites", result.get("Product1").type());
        Assert.assertEquals("value1", result.get("Product1").tags().get("tag1"));
        Assert.assertEquals("value3", result.get("Product1").tags().get("tag2"));
        // Resource 2
        Assert.assertEquals("2", result.get("Product2").id());
        Assert.assertEquals("Resource2", result.get("Product2").name());
        Assert.assertEquals("Building 44", result.get("Product2").location());
        // Resource 3
        Assert.assertEquals("3", result.get("Product3").id());
        Assert.assertEquals("Resource3", result.get("Product3").name());
    }

    @Test
    public void putDictionary() throws Exception {
        Map<String, FlattenedProduct> body = new HashMap<>();
        FlattenedProduct product = new FlattenedProduct();
        product.withLocation("West US");
        product.withTags(new HashMap<String, String>());
        product.tags().put("tag1", "value1");
        product.tags().put("tag2", "value3");
        product.withPname("Product1");
        product.withFlattenedProductType("Flat");
        body.put("Resource1", product);
        FlattenedProduct product1 = new FlattenedProduct();
        product1.withLocation("Building 44");
        product1.withPname("Product2");
        product1.withFlattenedProductType("Flat");
        body.put("Resource2", product1);
        client.putDictionary(body);
    }

    @Test
    public void getResourceCollection() throws Exception {
        ResourceCollection resultResource = client.getResourceCollection();
        //Dictionaryofresources
        Assert.assertEquals(3, resultResource.dictionaryofresources().size());
        // Resource 1
        Assert.assertEquals("1", resultResource.dictionaryofresources().get("Product1").id());
        Assert.assertEquals("OK", resultResource.dictionaryofresources().get("Product1").provisioningStateValues());
        Assert.assertEquals("Product1", resultResource.dictionaryofresources().get("Product1").pname());
        Assert.assertEquals("Flat", resultResource.dictionaryofresources().get("Product1").flattenedProductType());
        Assert.assertEquals("Building 44", resultResource.dictionaryofresources().get("Product1").location());
        Assert.assertEquals("Resource1", resultResource.dictionaryofresources().get("Product1").name());
        Assert.assertEquals("Succeeded", resultResource.dictionaryofresources().get("Product1").provisioningState());
        Assert.assertEquals("Microsoft.Web/sites", resultResource.dictionaryofresources().get("Product1").type());
        Assert.assertEquals("value1", resultResource.dictionaryofresources().get("Product1").tags().get("tag1"));
        Assert.assertEquals("value3", resultResource.dictionaryofresources().get("Product1").tags().get("tag2"));
        // Resource 2
        Assert.assertEquals("2", resultResource.dictionaryofresources().get("Product2").id());
        Assert.assertEquals("Resource2", resultResource.dictionaryofresources().get("Product2").name());
        Assert.assertEquals("Building 44", resultResource.dictionaryofresources().get("Product2").location());
        // Resource 3
        Assert.assertEquals("3", resultResource.dictionaryofresources().get("Product3").id());
        Assert.assertEquals("Resource3", resultResource.dictionaryofresources().get("Product3").name());

        //Arrayofresources
        Assert.assertEquals(3, resultResource.arrayofresources().size());
        // Resource 1
        Assert.assertEquals("4", resultResource.arrayofresources().get(0).id());
        Assert.assertEquals("OK", resultResource.arrayofresources().get(0).provisioningStateValues());
        Assert.assertEquals("Product4", resultResource.arrayofresources().get(0).pname());
        Assert.assertEquals("Flat", resultResource.arrayofresources().get(0).flattenedProductType());
        Assert.assertEquals("Building 44", resultResource.arrayofresources().get(0).location());
        Assert.assertEquals("Resource4", resultResource.arrayofresources().get(0).name());
        Assert.assertEquals("Succeeded", resultResource.arrayofresources().get(0).provisioningState());
        Assert.assertEquals("Microsoft.Web/sites", resultResource.arrayofresources().get(0).type());
        Assert.assertEquals("value1", resultResource.arrayofresources().get(0).tags().get("tag1"));
        Assert.assertEquals("value3", resultResource.arrayofresources().get(0).tags().get("tag2"));
        // Resource 2
        Assert.assertEquals("5", resultResource.arrayofresources().get(1).id());
        Assert.assertEquals("Resource5", resultResource.arrayofresources().get(1).name());
        Assert.assertEquals("Building 44", resultResource.arrayofresources().get(1).location());
        // Resource 3
        Assert.assertEquals("6", resultResource.arrayofresources().get(2).id());
        Assert.assertEquals("Resource6", resultResource.arrayofresources().get(2).name());

        //productresource
        Assert.assertEquals("7", resultResource.productresource().id());
        Assert.assertEquals("Resource7", resultResource.productresource().name());
    }

    @Test
    public void putResourceCollection() throws Exception {
        Map<String, FlattenedProduct> resources = new HashMap<>();
        resources.put("Resource1", new FlattenedProduct());
        resources.get("Resource1").withLocation("West US");
        resources.get("Resource1").withPname("Product1");
        resources.get("Resource1").withFlattenedProductType("Flat");
        resources.get("Resource1").withTags(new HashMap<String, String>());
        resources.get("Resource1").tags().put("tag1", "value1");
        resources.get("Resource1").tags().put("tag2", "value3");

        resources.put("Resource2", new FlattenedProduct());
        resources.get("Resource2").withLocation("Building 44");
        resources.get("Resource2").withPname("Product2");
        resources.get("Resource2").withFlattenedProductType("Flat");

        ResourceCollection complexObj = new ResourceCollection();
        complexObj.withDictionaryofresources(resources);
        complexObj.withArrayofresources(new ArrayList<FlattenedProduct>());
        complexObj.arrayofresources().add(resources.get("Resource1"));
        FlattenedProduct p1 = new FlattenedProduct();
        p1.withLocation("East US");
        p1.withPname("Product2");
        p1.withFlattenedProductType("Flat");
        complexObj.arrayofresources().add(p1);
        FlattenedProduct pr = new FlattenedProduct();
        pr.withLocation("India");
        pr.withPname("Azure");
        pr.withFlattenedProductType("Flat");
        complexObj.withProductresource(pr);

        client.putResourceCollection(complexObj);
    }

    @Test
    public void putSimpleProduct() throws Exception {
        SimpleProduct simpleProduct = new SimpleProduct();
        simpleProduct.withDescription("product description");
        simpleProduct.withProductId("123");
        simpleProduct.withMaxProductDisplayName("max name");
        simpleProduct.withCapacity("Large");
        simpleProduct.withOdatavalue("http://foo");
        simpleProduct.withGenericValue("https://generic");

        SimpleProduct product = client.putSimpleProduct(simpleProduct);
        assertSimpleProductEquals(simpleProduct, product);
    }

    @Test
    public void postFlattenedSimpleProduct() throws Exception {
        SimpleProduct simpleProduct = new SimpleProduct();
        simpleProduct.withDescription("product description");
        simpleProduct.withProductId("123");
        simpleProduct.withMaxProductDisplayName("max name");
        simpleProduct.withCapacity("Large");
        simpleProduct.withOdatavalue("http://foo");
        client.postFlattenedSimpleProduct("123", "max name", "product description", null, "http://foo");
    }

    @Test
    public void putSimpleProductWithGrouping() throws Exception {
        SimpleProduct simpleProduct = new SimpleProduct();
        simpleProduct.withDescription("product description");
        simpleProduct.withProductId("123");
        simpleProduct.withMaxProductDisplayName("max name");
        simpleProduct.withCapacity("Large");
        simpleProduct.withOdatavalue("http://foo");

        FlattenParameterGroup flattenParameterGroup = new FlattenParameterGroup();
        flattenParameterGroup.withDescription("product description");
        flattenParameterGroup.withProductId("123");
        flattenParameterGroup.withMaxProductDisplayName("max name");
        flattenParameterGroup.withOdatavalue("http://foo");
        flattenParameterGroup.withName("groupproduct");

        SimpleProduct product = client.putSimpleProductWithGrouping(flattenParameterGroup);
        assertSimpleProductEquals(simpleProduct, product);
    }

    private void assertSimpleProductEquals(SimpleProduct expected, SimpleProduct actual) throws Exception {
        Assert.assertEquals(expected.productId(), actual.productId());
        Assert.assertEquals(expected.description(), actual.description());
        Assert.assertEquals(expected.capacity(), actual.capacity());
        Assert.assertEquals(expected.maxProductDisplayName(), actual.maxProductDisplayName());
        Assert.assertEquals(expected.odatavalue(), actual.odatavalue());
    }
}

File: src/generator/AutoRest.Python.Tests/AcceptanceTests/model_flattening_tests.py
# --------------------------------------------------------------------------
#
# Copyright (c) Microsoft Corporation. All rights reserved.
#
# The MIT License (MIT)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the ""Software""), to
# deal in the Software without restriction, including without limitation the
# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
# sell copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
# IN THE SOFTWARE.
#
# --------------------------------------------------------------------------

import unittest
import subprocess
import sys
import isodate
import tempfile
import json
from datetime import date, datetime, timedelta
import os
from os.path import dirname, pardir, join, realpath

cwd = dirname(realpath(__file__))
log_level = int(os.environ.get('PythonLogLevel', 30))

tests = realpath(join(cwd, pardir, "Expected", "AcceptanceTests"))
sys.path.append(join(tests, "ModelFlattening"))

from msrest.serialization import Deserializer
from msrest.exceptions import DeserializationError

from auto_rest_resource_flattening_test_service import AutoRestResourceFlatteningTestService
from auto_rest_resource_flattening_test_service.models import (
    FlattenedProduct,
    ErrorException,
    ResourceCollection,
    SimpleProduct,
    FlattenParameterGroup)

class ModelFlatteningTests(unittest.TestCase):

    def setUp(self):
        self.client = AutoRestResourceFlatteningTestService(base_url="http://localhost:3000")
        return super(ModelFlatteningTests, self).setUp()

    def test_flattening_array(self):

        #Array
        result = self.client.get_array()
        self.assertEqual(3, len(result))
        # Resource 1
        self.assertEqual("1", result[0].id)
        self.assertEqual("OK", result[0].provisioning_state_values)
        self.assertEqual("Product1", result[0].pname)
        self.assertEqual("Flat", result[0].flattened_product_type)
        self.assertEqual("Building 44", result[0].location)
        self.assertEqual("Resource1", result[0].name)
        self.assertEqual("Succeeded", result[0].provisioning_state)
        self.assertEqual("Microsoft.Web/sites", result[0].type)
        self.assertEqual("value1", result[0].tags["tag1"])
        self.assertEqual("value3", result[0].tags["tag2"])
        # Resource 2
        self.assertEqual("2", result[1].id)
        self.assertEqual("Resource2", result[1].name)
        self.assertEqual("Building 44", result[1].location)
        # Resource 3
        self.assertEqual("3", result[2].id)
        self.assertEqual("Resource3", result[2].name)

        resourceArray = [
                {
                    'location': "West US",
                    'tags': {"tag1":"value1", "tag2":"value3"}},
                {
                    'location': "Building 44"}]

        self.client.put_array(resourceArray)

    def test_flattening_dictionary(self):

        #Dictionary
        resultDictionary = self.client.get_dictionary()
        self.assertEqual(3, len(resultDictionary))
        # Resource 1
        self.assertEqual("1", resultDictionary["Product1"].id)
        self.assertEqual("OK", resultDictionary["Product1"].provisioning_state_values)
        self.assertEqual("Product1", resultDictionary["Product1"].pname)
        self.assertEqual("Flat", resultDictionary["Product1"].flattened_product_type)
        self.assertEqual("Building 44", resultDictionary["Product1"].location)
        self.assertEqual("Resource1", resultDictionary["Product1"].name)
        self.assertEqual("Succeeded", resultDictionary["Product1"].provisioning_state)
        self.assertEqual("Microsoft.Web/sites", resultDictionary["Product1"].type)
        self.assertEqual("value1", resultDictionary["Product1"].tags["tag1"])
        self.assertEqual("value3", resultDictionary["Product1"].tags["tag2"])
        # Resource 2
        self.assertEqual("2", resultDictionary["Product2"].id)
        self.assertEqual("Resource2", resultDictionary["Product2"].name)
        self.assertEqual("Building 44", resultDictionary["Product2"].location)
        # Resource 3
        self.assertEqual("3", resultDictionary["Product3"].id)
        self.assertEqual("Resource3", resultDictionary["Product3"].name)

        resourceDictionary = {
                "Resource1": {
                    'location': "West US",
                    'tags': {"tag1":"value1", "tag2":"value3"},
                    'pname': "Product1",
                    'flattened_product_type': "Flat"},
                "Resource2": {
                    'location': "Building 44",
                    'pname': "Product2",
                    'flattened_product_type': "Flat"}}

        self.client.put_dictionary(resourceDictionary)

    def test_flattening_complex_object(self):

        #ResourceCollection
        resultResource = self.client.get_resource_collection()

        #dictionaryofresources
        self.assertEqual(3, len(resultResource.dictionaryofresources))
        # Resource 1
        self.assertEqual("1", resultResource.dictionaryofresources["Product1"].id)
        self.assertEqual("OK", resultResource.dictionaryofresources["Product1"].provisioning_state_values)
        self.assertEqual("Product1", resultResource.dictionaryofresources["Product1"].pname)
        self.assertEqual("Flat", resultResource.dictionaryofresources["Product1"].flattened_product_type)
        self.assertEqual("Building 44", resultResource.dictionaryofresources["Product1"].location)
        self.assertEqual("Resource1", resultResource.dictionaryofresources["Product1"].name)
        self.assertEqual("Succeeded", resultResource.dictionaryofresources["Product1"].provisioning_state)
        self.assertEqual("Microsoft.Web/sites", resultResource.dictionaryofresources["Product1"].type)
        self.assertEqual("value1", resultResource.dictionaryofresources["Product1"].tags["tag1"])
        self.assertEqual("value3", resultResource.dictionaryofresources["Product1"].tags["tag2"])
        # Resource 2
        self.assertEqual("2", resultResource.dictionaryofresources["Product2"].id)
        self.assertEqual("Resource2", resultResource.dictionaryofresources["Product2"].name)
        self.assertEqual("Building 44", resultResource.dictionaryofresources["Product2"].location)
        # Resource 3
        self.assertEqual("3", resultResource.dictionaryofresources["Product3"].id)
        self.assertEqual("Resource3", resultResource.dictionaryofresources["Product3"].name)

        #arrayofresources
        self.assertEqual(3, len(resultResource.arrayofresources))
        # Resource 1
        self.assertEqual("4", resultResource.arrayofresources[0].id)
        self.assertEqual("OK", resultResource.arrayofresources[0].provisioning_state_values)
        self.assertEqual("Product4", resultResource.arrayofresources[0].pname)
        self.assertEqual("Flat", resultResource.arrayofresources[0].flattened_product_type)
        self.assertEqual("Building 44", resultResource.arrayofresources[0].location)
        self.assertEqual("Resource4", resultResource.arrayofresources[0].name)
        self.assertEqual("Succeeded", resultResource.arrayofresources[0].provisioning_state)
        self.assertEqual("Microsoft.Web/sites", resultResource.arrayofresources[0].type)
        self.assertEqual("value1", resultResource.arrayofresources[0].tags["tag1"])
        self.assertEqual("value3", resultResource.arrayofresources[0].tags["tag2"])
        # Resource 2
        self.assertEqual("5", resultResource.arrayofresources[1].id)
        self.assertEqual("Resource5", resultResource.arrayofresources[1].name)
        self.assertEqual("Building 44", resultResource.arrayofresources[1].location)
        # Resource 3
        self.assertEqual("6", resultResource.arrayofresources[2].id)
        self.assertEqual("Resource6", resultResource.arrayofresources[2].name)

        #productresource
        self.assertEqual("7", resultResource.productresource.id)
        self.assertEqual("Resource7", resultResource.productresource.name)

        resourceDictionary = {
                "Resource1": FlattenedProduct(
                    location = "West US",
                    tags = {"tag1":"value1", "tag2":"value3"},
                    pname = "Product1",
                    flattened_product_type = "Flat"),
                "Resource2": FlattenedProduct(
                    location = "Building 44",
                    pname = "Product2",
                    flattened_product_type = "Flat")}

        resourceComplexObject = ResourceCollection(
                dictionaryofresources = resourceDictionary,
                arrayofresources = [
                    FlattenedProduct(
                        location = "West US",
                        tags = {"tag1":"value1", "tag2":"value3"},
                        pname = "Product1",
                        flattened_product_type = "Flat"),
                    FlattenedProduct(
                        location = "East US",
                        pname = "Product2",
                        flattened_product_type = "Flat")],
                productresource = FlattenedProduct(
                    location = "India",
                    pname = "Azure",
                    flattened_product_type = "Flat"))

        self.client.put_resource_collection(resourceComplexObject)

    def test_model_flattening_simple(self):

        simple_prduct = SimpleProduct(
            product_id = "123",
            description = "product description",
            max_product_display_name = "max name",
            odatavalue = "http://foo",
            generic_value = "https://generic"
            )

        result = self.client.put_simple_product(simple_prduct)
        self.assertEqual(result, simple_prduct)

    def test_model_flattening_with_parameter_flattening(self):

        simple_product = SimpleProduct(
            product_id = "123",
            description = "product description",
            max_product_display_name = "max name",
            odatavalue = "http://foo"
            )

        result = self.client.post_flattened_simple_product("123", "max name", "product description", None, "http://foo")
        self.assertEqual(result, simple_product)

    def test_model_flattening_with_grouping(self):

        simple_prduct = SimpleProduct(
            product_id = "123",
            description = "product description",
            max_product_display_name = "max name",
            odatavalue = "http://foo"
            )

        group = FlattenParameterGroup(
            product_id = "123",
            description = "product description",
            max_product_display_name="max name",
            odatavalue="http://foo",
            name="groupproduct")

        result = self.client.put_simple_product_with_grouping(group)
        self.assertEqual(result, simple_prduct)

if __name__ == '__main__':
    unittest.main()
